agent/lib/kontena/workers/volumes/volume_manager.rb

```diff
@@ -94,10 +97,20 @@ def sync_volume_to_master(docker_volume)
       rpc_client.async.request('/node_volumes/set_state', [node.id, volume])
     end
 
-    def volume_exist?(volume_name)
+    # Checks if given volume exists with the expected driver
+    #
+    # @param [String] name of the volume
+    # @param [String] driver to expect on the volume if already existing
+    # @raise [DriverMismatchError] If the volume is found but using a different driver than expected
+    def volume_exist?(volume_name, driver)
       begin
         debug "volume #{volume_name} exists"
-        true if Docker::Volume.get(volume_name)
+        volume = Docker::Volume.get(volume_name)
+        if volume && volume.info['Driver'] == driver
+          return true
+        elsif volume && volume.info['Driver'] != driver
+          raise DriverMismatchError.new("Volume driver not as expected. Expected #{driver}, existing volume has #{volume.info['Driver']}")
```

Comment:
I think the `ensure_volume` would be the right place to handle this error... change it to call `volume_exist?` and log it?

Then the agent will immediately warn about these volume conflicts earlier, and we could even later extend it with some kind of error reporting, and perhaps even have the scheduler avoid this node or something?